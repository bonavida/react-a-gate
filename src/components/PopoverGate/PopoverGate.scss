.popover {
  &__container {
    position: absolute;
    display: inline-block;

    &--top {
      margin-top: -5px;
    }

    &--bottom {
      margin-top: 5px;
    }

    &--left {
      margin-left: -5px;
    }

    &--right {
      margin-left: 5px;
    }
  }

  &__arrow {
    position: absolute;
    width: 0;
    height: 0;
    border-style: solid;
    border-color: #222;
    z-index: 3;

    &--top {
      margin-top: -6px;
      border-width: 5px 5px 0;
      border-bottom-color: transparent !important;
      border-left-color: transparent !important;
      border-right-color: transparent !important;
    }

    &--bottom {
      margin-top: 6px;
      border-width: 0 5px 5px;
      border-top-color: transparent !important;
      border-left-color: transparent !important;
      border-right-color: transparent !important;
    }

    &--left {
      margin-left: -6px;
      border-width: 5px 0 5px 5px;
      border-top-color: transparent !important;
      border-bottom-color: transparent !important;
      border-right-color: transparent !important;
    }

    &--right {
      margin-left: 6px;
      border-width: 5px 5px 5px 0;
      border-top-color: transparent !important;
      border-bottom-color: transparent !important;
      border-left-color: transparent !important;
    }
  }

  &__arrow-border {
    border-color: #666;
    z-index: 2;

    &.popover__arrow {
      &--top {
        margin-left: -1px;
        border-width: 6px 6px 0;
      }
      &--bottom {
        margin-top: 5px;
        margin-left: -1px;
        border-width: 0 6px 6px;
      }
      &--left {
        margin-top: -1px;
        border-width: 6px 0 6px 6px;
      }
      &--right {
        margin-left: 5px;
        margin-top: -1px;
        border-width: 6px 6px 6px 0;
      }
    }
  }

  &__inner {
    max-width: 200px;
    padding: 0.3rem 0.5rem;
    color: #fff;
    font-size: 0.8em;
    text-align: center;
    background-color: #222;
    border: 1px solid #666;
    border-radius: 0.25rem;
    box-shadow: 0 1px 10px rgba(0, 0, 0, 0.3);
  }
}

.light {
  .popover__inner {
    background-color: #fff;
    border: 1px solid #e1e4e8;
    color: #333;
    box-shadow: 0 8px 24px rgba(149, 157, 165, 0.2);
  }

  .popover__arrow {
    border-color: #fff;
  }

  .popover__arrow-border {
    border-color: rgba(27, 31, 35, 0.13);
  }
}
