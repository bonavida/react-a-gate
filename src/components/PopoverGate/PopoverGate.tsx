import * as React from 'react';
import {
  useState,
  useEffect,
  useLayoutEffect,
  useRef,
  useCallback,
} from 'react';
import { createPortal } from 'react-dom';
/** Hooks */
import usePortal from 'hooks/usePortal';
import useResizeObserver from 'hooks/useResizeObserver';
/** Utils */
import { calculatePosition } from 'utils';
/** Types */
import { Place, Mode, Theme } from 'types';
/** Styles */
import './PopoverGate.scss';

export type PopoverGateProps = {
  rootId?: string;
  className?: string;
  content: string | JSX.Element;
  place?: Place;
  mode?: Mode;
  closeWhenClickOutside?: boolean;
  theme?: Theme;
  offset?: number;
};

const PopoverGate: React.FC<PopoverGateProps> = ({
  rootId = 'portal-root',
  className,
  content,
  place = 'top',
  mode = 'click',
  closeWhenClickOutside = true,
  theme,
  offset = 10,
  children,
}) => {
  const [showPopover, setShowPopover] = useState(false);
  const [coords, setCoords] = useState({ top: -99999, left: -99999 });
  const [arrowCoords, setArrowCords] = useState({ top: -99999, left: -99999 });
  const [placePopover, setPlacePopover] = useState(place);
  const popoverRef = useRef<HTMLDivElement>(null);
  const arrowRef = useRef<HTMLDivElement>(null);
  const triggerRef = useRef<JSX.Element>(null);
  const target = usePortal(rootId);

  const handleCloseWhenClickOutside = useCallback(
    (e) => {
      if (
        closeWhenClickOutside &&
        triggerRef.current !== e.target &&
        !popoverRef.current?.contains(e.target)
      ) {
        setShowPopover(false);
      }
    },
    [closeWhenClickOutside]
  );

  const updatePosition = useCallback(() => {
    if (triggerRef.current) {
      const {
        top,
        left,
        arrowTop,
        arrowLeft,
        place: newPlace,
      } = calculatePosition(triggerRef, popoverRef, arrowRef, place, offset);
      setCoords({ top, left });
      setArrowCords({ top: arrowTop, left: arrowLeft });
      setPlacePopover(newPlace);
    }
  }, [place, offset]);

  useResizeObserver(document.body, updatePosition);

  useLayoutEffect(() => {
    if (popoverRef.current) {
      updatePosition();
    }
  }, [showPopover, updatePosition]);

  useEffect(() => {
    if (mode === 'click') {
      window.addEventListener('click', handleCloseWhenClickOutside);
    }

    return () => {
      if (mode === 'click') {
        window.removeEventListener('click', handleCloseWhenClickOutside);
      }
    };
  }, [mode, handleCloseWhenClickOutside]);

  const popoverEvents = useCallback(() => {
    switch (mode) {
      case 'click':
      default:
        return {
          onClick: () => {
            setShowPopover((showPopoverState) => !showPopoverState);
          },
        };
      case 'hover':
        return {
          onMouseEnter: () => {
            setShowPopover(true);
          },
          onMouseLeave: () => {
            setShowPopover(false);
          },
        };
    }
  }, [mode]);

  return (
    <>
      {React.cloneElement(children as JSX.Element, {
        ...popoverEvents(),
        'data-testid': 'popover_trigger',
        ref: triggerRef,
      })}
      {showPopover &&
        createPortal(
          <div className={className || theme || 'popover'}>
            <div
              ref={popoverRef}
              className={`popover__container popover__container--${placePopover}`}
              style={{ ...coords }}
              role="tooltip"
            >
              <div className="popover__inner">{content}</div>
            </div>
            <div
              ref={arrowRef}
              className={`popover__arrow popover__arrow--${placePopover}`}
              style={{ ...arrowCoords }}
            />
            <div
              className={`tooltip__arrow tooltip__arrow-border tooltip__arrow--${placePopover}`}
              style={{ ...arrowCoords }}
            />
          </div>,
          target
        )}
    </>
  );
};

export default PopoverGate;
