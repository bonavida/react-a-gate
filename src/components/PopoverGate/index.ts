import PopoverGate from './PopoverGate';
export * from './PopoverGate';

export default PopoverGate;
