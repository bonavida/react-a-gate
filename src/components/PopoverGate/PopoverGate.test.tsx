import * as React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
/** Mocks */
import ResizeObserverMock from '../../mocks/ResizeObserver';
/** Components and types */
import PopoverGate, { PopoverGateProps } from './PopoverGate';

afterEach(cleanup);

describe('PopoverGate', () => {
  let props: PopoverGateProps;

  beforeEach(() => {
    props = { content: 'popover text' };
    window.ResizeObserver = ResizeObserverMock;
  });

  it('should match snapshot', () => {
    // Arrange
    const { asFragment } = render(
      <PopoverGate {...props}>
        <span>test</span>
      </PopoverGate>
    );

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display trigger content', () => {
    // Arrange
    const { queryByText } = render(
      <PopoverGate {...props}>
        <span>test</span>
      </PopoverGate>
    );

    // Assert
    expect(queryByText('test')).toBeTruthy();
  });

  it('should not display popover content if trigger element is not being clicked (when mode = click)', () => {
    // Arrange
    const { queryByText } = render(
      <PopoverGate {...props}>
        <span>test</span>
      </PopoverGate>
    );

    // Assert
    expect(queryByText('popover')).toBeFalsy();
  });

  it('should display popover content if trigger element is being clicked (when mode = click)', () => {
    // Arrange
    const { getByText, getByTestId } = render(
      <PopoverGate {...props}>
        <span>test</span>
      </PopoverGate>
    );

    fireEvent.click(getByTestId('popover_trigger'));

    // Assert
    expect(getByText(/popover/i)).toBeTruthy();
  });

  it('should not display popover content if trigger element is not being hovered (when mode = hover)', () => {
    // Arrange
    const { queryByText } = render(
      <PopoverGate mode="hover" {...props}>
        <span>test</span>
      </PopoverGate>
    );

    // Assert
    expect(queryByText('popover')).toBeFalsy();
  });

  it('should display popover content if trigger element is being hovered (when mode = hover)', () => {
    // Arrange
    const { getByText, getByTestId } = render(
      <PopoverGate mode="hover" {...props}>
        <span>test</span>
      </PopoverGate>
    );

    fireEvent.mouseEnter(getByTestId('popover_trigger'));

    // Assert
    expect(getByText(/popover/i)).toBeTruthy();
  });

  it("should remain opened if it's clicked inside the popover content", () => {
    // Act
    const { getByText } = render(
      <PopoverGate {...props}>
        <span>test</span>
      </PopoverGate>
    );

    fireEvent.click(getByText(/test/i));

    // Assert
    expect(getByText(/popover/i)).toBeTruthy();
  });

  it("should be closed if it's clicked outside the popover content", () => {
    // Act
    const { queryByText } = render(
      <PopoverGate {...props}>
        <span>test</span>
      </PopoverGate>
    );

    fireEvent.click(document);

    // Assert
    expect(queryByText(/popover/i)).toBeFalsy();
  });
});
