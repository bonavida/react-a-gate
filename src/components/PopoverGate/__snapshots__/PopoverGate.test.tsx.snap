// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverGate should match snapshot 1`] = `
<DocumentFragment>
  <span
    data-testid="popover_trigger"
  >
    test
  </span>
</DocumentFragment>
`;
